nearcore/src/download_file.rs

```diff
@@ -0,0 +1,369 @@
+    /// when performing decompression.  Calling it with an empty `chunk`
+    /// indicates the end of the compressed data.
+    async fn decompress(
+        path: &std::path::Path,
+        file: &mut tokio::fs::File,
+        stream: &mut xz2::stream::Stream,
+        buffer: &mut [u8],
+        mut chunk: &[u8],
+    ) -> Result<(), FileDownloadError> {
+        let action =
+            if chunk.is_empty() { xz2::stream::Action::Finish } else { xz2::stream::Action::Run };
+        loop {
+            let total_in = stream.total_in();
+            let total_out = stream.total_out();
+            let status = stream.process(chunk, buffer, action)?;
+            match status {
+                xz2::stream::Status::Ok => (),
+                xz2::stream::Status::StreamEnd => (),
+                status => {
+                    let status = format!("{:?}", status);
+                    tracing::error!(target: "near", "Got unexpected status ‘{}’ when decompressing downloaded file.", status);
+                    return Err(FileDownloadError::XzStatusError(status));
+                }
+            };
+            let read = (stream.total_in() - total_in).try_into().unwrap();
+            chunk = &chunk[read..];
+            let out = (stream.total_out() - total_out).try_into().unwrap();
+            file.write_all(&buffer[..out])
+                .await
+                .map_err(|e| FileDownloadError::WriteError(path.to_path_buf(), e))?;
+            if chunk.is_empty() {
+                break Ok(());
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hyper::service::{make_service_fn, service_fn};
+    use hyper::{Body, Request, Response, Server};
+    use std::convert::Infallible;
+
+    #[tokio::test]
+    async fn test_file_download() {
+        let listener = std::net::TcpListener::bind("127.0.0.1:0").unwrap();
+        let port = listener.local_addr().unwrap().port();
+
+        async fn handle_request(_req: Request<Body>) -> Result<Response<Body>, Infallible> {
+            let data: [u8; 1024] = [42; 1024];
+            Ok(Response::new(Body::from(data.to_vec())))
+        }
+
+        tokio::task::spawn(async move {
+            let make_svc =
+                make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(handle_request)) });
+            let server = Server::from_tcp(listener).unwrap().serve(make_svc);
+            if let Err(e) = server.await {
+                eprintln!("server error: {}", e);
+            }
+        });
+
+        let tmp_downloaded_file = tempfile::NamedTempFile::new().unwrap();
+        let tmp_downloaded_file_path = tmp_downloaded_file.path();
+
+        download_file(&format!("http://localhost:{}", port), tmp_downloaded_file_path)
+            .await
+            .unwrap();
+
+        let downloaded_file_content = std::fs::read(tmp_downloaded_file_path).unwrap();
+        assert_eq!(downloaded_file_content, [42; 1024].to_vec());
+    }
```

Comment:
How about converting it into a helper function with `expected` and `payload` arguments and than running the test once for plain data and another time for XZ-compressed data?

Reply:
I've thought about this, but I think it's better not to do this. The way tests are written, we have control over decoder's state machine (`for chunk_size` loops). If we hit it purely via downloading, than I think we'll only handle the boring case where all the data is available up-front.

Reply:
After thinking for one more minute, I realized that it's probably better to add the tests you suggest, but keep the old ones as well :-)